CNV-40785: Configure guest for real-time (#75714)
openshift/openshift-docs: modules/virt-configuring-vm-real-time.adoc

// Module included in the following assemblies:
//
// * virt/virtual_machines/advanced_vm_management/virt-configuring-cluster-realtime-workloads.adoc

> 

Reviewer: Need to add the empty line back

:_mod-docs-content-type: PROCEDURE
[id="virt-configuring-vm-real-time_{context}"]
= Configuring a virtual machine for real-time workloads

> 

Reviewer: Same here (add empty line)

> You can configure a virtual machine (VM) to run real-time workloads.

Reviewer: Need empty line below

You can configure a virtual machines (VM) to run real-time workloads.

You can configure a virtual machine (VM) to run real-time workloads.
.Prerequisites
* Your cluster is configured to run real-time workloads.
* You have installed the `virtctl` tool.

> 

Reviewer: Add back the empty line before the `.Procedure` heading

>   template:

Author: Done.

<6> The size of the huge pages. The possible values for x86-64 architectures are `1Gi` and `2Mi`. In this example, the request is for 4 huge pages of size 1 Gi.
. Apply the `VirtualMachine` manifest:
+
[source,terminal]
----
$ oc apply -f <file_name>.yaml
----
. Configure the guest operating system. The following example shows the configuration steps for a {op-system-base} 9 operating system:
.. Run the following command to connect to the VM console:
+
[source,terminal]
----
$ virtctl console <vm_name>
----
.. Disable the `irqbalance` service by using the following command:
+
[source,terminal]
----
$ systemctl disable irqbalance && systemctl stop irqbalance
----
.. Install the necessary packages by running the following command:
+
[source,terminal]
----
$ dnf install kernel-rt tuned cloud-init realtime-tests
----
.. To achieve low-latency tuning by using the `realtime-virtual-guest` profile in the TuneD application, run the following commands:
+
[source,terminal]
----
$ dnf --enablerepo=rt install -y tuned-profiles-realtime
----
+
[source,terminal]
----
$ dnf --enablerepo=nfv install -y tuned-profiles-nfv-guest
----
.. Edit `/etc/tuned/realtime-virtual-guest-variables.conf`:
+

> [source,terminal]

Reviewer: ```suggestion
[source,conf]
```

# managed IRQs out of isolated cores. Note that this requires kernel
# support. Please only specify this parameter if you are sure that the
# kernel supports it.
#
isolate_managed_irq=Y <2>

#
# Set the desired combined queue count value using the parameter provided
# below. Ideally this should be set to the number of housekeeping CPUs i.e.,
# in the example given below it is assumed that the system has 4 housekeeping
# (non-isolated) CPUs.
#
# netdev_queue_count=4
----
<1> The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.
<2> isolate_managed_irq should be set to `Y`
.. Activate the TuneD profile:
+
[source,terminal]
----
$ tuned-adm profile realtime-virtual-guest
----
.. Set the real-time kernel as the default by using the GRUB boot loader command-line interface:
+
[source,terminal]
----
$ grubby --set-default=$(ls /boot/vmlinuz* | grep rt)
----
.. Set the kernel arguments using the GRUB boot loader command-line interface:
+
[source,terminal]
----
$ grubby --args="iommu=pt intel_iommu=on default_hugepagesz=1G idle=poll" --update-kernel=$(grubby --default-kernel)
----
. Restart the VM to apply the changes.

> .Verification

Reviewer: Add empty line above the `.Verification` heading

        memory:
          guest: 4Gi
          hugepages:
            pageSize: 1Gi # <6>
        terminationGracePeriodSeconds: 0

> # ...

Reviewer 2: As discussed - we need a more explicit VM manifest example, that perfectly suits realtime. It must include the `image` entry, because a realtime VM is not a standard one, and the user might not have created and pushed their own image.

Author: @phoracek @sjhala-ccs WDYT?

Author: BTW, the example DPDK VM spec does not have a disk as well https://docs.openshift.com/container-platform/4.15/virt/vm_networking/virt-using-dpdk-with-sriov.html#virt-configuring-vm-dpdk_virt-using-dpdk-with-sriov

Reviewer 3: This is not the only place where we leave out part of the VM yaml. I think it makes sense to document only the "special" parts, to make this easier to maintaine by making if future-proof, and also put the spotlight on the important attributes.

That being said, I will not fight against having a full example either.

<2> This annotation specifies that the CPU quota is disabled for CPUs that are used by the container.
<3> This annotation specifies that interrupt request (IRQ) load balancing is disabled for CPUs that are used by the container.
<4> The number of sockets inside the VM.
<5> The number of cores inside the VM. This must be a value greater than or equal to `1`.
<6> The size of the huge pages. The possible values for x86-64 architectures are `1Gi` and `2Mi`. In this example, the request is for 4 huge pages of size 1 Gi.
. Apply the `VirtualMachine` manifest:
+
[source,terminal]
----
$ oc apply -f <file_name>.yaml
----
. Configure the guest operating system. The following example shows the configuration steps for a {op-system-base} 9 operating system:
.. Run the following command to connect to the VM console:
+
[source,terminal]
----
$ virtctl console <vm_name>
----
.. Disable the `irqbalance` service by using the following command:
+
[source,terminal]
----
$ systemctl disable irqbalance && systemctl stop irqbalance
----
.. Install the necessary packages by running the following command:
+
[source,terminal]
----

> $ dnf install kernel-rt tuned cloud-init realtime-tests

Author: Have you tested it with RHEL 9?
Please see @ngu-niny' [comment](https://github.com/openshift/openshift-docs/pull/75714#discussion_r1597832897).

> $ dnf install kernel-rt tuned cloud-init realtime-tests

Reviewer 2: Following this comment, I tried a RHEL9 image which I took from `registry.redhat.io/rhel9/rhel-guest-image`.
The result was 
```No match for argument: kernel-rt
Error: Unable to find a match: kernel-rt
```
(the other packages were either installed successfully, or were already present).
So I tried following the instructions in https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux_for_real_time/9/html-single/installing_rhel_9_for_real_time/index (after registering to subscription-manager), and the result was
```
$ sudo subscription-manager repos --enable rhel-9-for-x86_64-rt-rpms
Error: 'rhel-9-for-x86_64-rt-rpms' does not match a valid repository ID. Use "subscription-manager repos --list" to see valid repositories.
```
Maybe I'm underestimating the advanced user who is supposed to be the audience for this feature, but I think there are too many obstacles on the way for running a successful realtime check. Maybe we should re-consider the importance of providing a RHEL 9 VM image.
@orelmisan @phoracek WDYT?

> $ dnf install kernel-rt tuned cloud-init realtime-tests

Reviewer 4: I used the KVM Guest Image in https://access.redhat.com/downloads/content/479/ver=/rhel---9/9.4/x86_64/product-software in the VM manifest, as following url part shows.

spec:
  dataVolumeTemplates:
  - apiVersion: cdi.kubevirt.io/v1beta1
    kind: DataVolume
    metadata:
      creationTimestamp: null
      name: rhel94-rt-5cpu-84-1
    spec:
      source:
        http:
          url: https://access.cdn.redhat.com/content/origin/files/sha256/d3/d362e72518a2d7415d850b8177c814f6fd87f42ca1640bda17e98855e1d6baad/rhel-9.4-x86_64-kvm.qcow2?user=3124af3237f60c89937f85f9f98896b7&_auth_=1716880954_459335f349cede3ac1e894c9c9b9acb5

Then in the guest, do following steps to enable the corresponding repos.
# subscription-manager register
# subscription-manager repos --enable rhel-9-for-x86_64-rt-rpms
# subscription-manager repos --enable rhel-9-for-x86_64-nfv-rpms

It works for most of the steps except following two commands:
# dnf --enablerepo=rt install -y tuned-profiles-realtime
# dnf --enablerepo=nfv install -y tuned-profiles-nfv-guest

After delete the '--enablerepo=*', it works:
# dnf install -y tuned-profiles-realtime
# dnf install -y tuned-profiles-nfv-guest

> $ dnf install kernel-rt tuned cloud-init realtime-tests

Reviewer 2: Eventually, Orel supplied a VM with DataVolume that managed to run successfully, apply all the additional commands and run the checkup successfully.
Thank you @orelmisan , and thank you as well @ngu-niny for your support here.

# isolated. By default we reserve 1 core per socket for housekeeping
# and isolate the rest. But you can isolate any range as shown in the
# examples below. Just remember to keep only one isolated_cores= line.
#
# Examples:
# isolated_cores=2,4-7
# isolated_cores=2-23
#
# Reserve 1 core per socket for housekeeping, isolate the rest.
# Change this for a core list or range as shown above.
isolated_cores=2-3 <1>

#
# Uncomment the 'isolate_managed_irq=Y' bellow if you want to move kernel
# managed IRQs out of isolated cores. Note that this requires kernel
# support. Please only specify this parameter if you are sure that the
# kernel supports it.
#
isolate_managed_irq=Y <2>

#
# Set the desired combined queue count value using the parameter provided
# below. Ideally this should be set to the number of housekeeping CPUs i.e.,
# in the example given below it is assumed that the system has 4 housekeeping
# (non-isolated) CPUs.
#
# netdev_queue_count=4
----
<1> The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.
<2> isolate_managed_irq should be set to `Y`
.. Activate the TuneD profile:
+
[source,terminal]
----

> $ tuned-adm profile realtime-virtual-guest

Reviewer 2: Also requires root.

----
.. Set the real-time kernel as the default by using the GRUB boot loader command-line interface:
+
[source,terminal]
----

> $ grubby --set-default=<installed_rt_kernel_version>

Reviewer 2: 1. I'm using the U/S image `quay.io/kiagnose/kubevirt-realtime-checkup-vm:v0.2.0`, which produces a CentOS VM.
The kernel version as retrieved by either `uname -r` or `hostnamectl` is invalid for this command:
```
[cloud-user@realtime-vm tuned]$ uname -r
4.18.0-348.rt7.130.el8.x86_64
[cloud-user@realtime-vm tuned]$ sudo grubby --set-default=4.18.0-348.rt7.130.el8.x86_64
The param 4.18.0-348.rt7.130.el8.x86_64 is incorrect
```

[UPDATE] The cyclictest passed successfully (i.e. `Max Latencies` < 40us) despite this, so @orelmisan @RamLavi - is this setting necessary?

2. Also requires root

Author: `quay.io/kiagnose/kubevirt-realtime-checkup-vm:v0.2.0` is based on CentOS stream 8 and the real-time checkup does additional configuration to it after it boots.
Please test it with a "clean" RHEL 9 installation.

----
.. Set the kernel arguments using the GRUB boot loader command-line interface:
+
[source,terminal]
----

> $ grubby --args="iommu=pt intel_iommu=on default_hugepagesz=1G idle=poll" --update-kernel=$(grubby --default-kernel)

Reviewer 2: root

----
. Restart the VM to apply the changes.

.Verification
* Use the `cyclictest` tool to verify that the real-time guest is configured properly:
+
[source,terminal]
----

> $ cyclictest --priority 1 --policy fifo -h 50 -a 2-3 --mainaffinity 0,1 -t 2 -m -q -i 200 -D 12h

Reviewer 2: 1. The user should be alerted that the `-D` parameter determines the test duration, so if they use this example it means a 12h test.
2. This also requires root

Author: @sjhala-ccs can we somehow refer to the man page of cyclictest?

Reviewer 2: > The user should be alerted that the -D parameter determines the test duration, so if they use this example it means a 12h test.

This should still be addressed.

        memory:
          guest: 4Gi
          hugepages:
            pageSize: 1Gi # <6>
        terminationGracePeriodSeconds: 0
# ...
----
<1> This annotation specifies that load balancing is disabled for CPUs that are used by the container.
<2> This annotation specifies that the CPU quota is disabled for CPUs that are used by the container.
<3> This annotation specifies that interrupt request (IRQ) load balancing is disabled for CPUs that are used by the container.
<4> The number of sockets inside the VM.
<5> The number of cores inside the VM. This must be a value greater than or equal to `1`.
<6> The size of the huge pages. The possible values for x86-64 architectures are `1Gi` and `2Mi`. In this example, the request is for 4 huge pages of size 1 Gi.
. Apply the `VirtualMachine` manifest:
+
[source,terminal]
----
$ oc apply -f <file_name>.yaml
----
. Configure the guest operating system. The following example shows the configuration steps for a {op-system-base} 9 operating system:

> .. Run the following command to connect to the VM console:

Reviewer: ```suggestion
.. Run the following command to connect to the VM console as a root user:
```

Author: This is not required (and is a potential security risk), since the command is run on the user's laptop for example.
The commands which are executed inside the guest, do required root privilages.

Reviewer: Understood. Then we can keep the wording and the command as is. I will mark the two comments related to this step as resolved.

+
[source,terminal]
----

> $ virtctl console <vm_name>

Reviewer: ```suggestion
$ sudo virtctl console <vm_name>
```

Author: This is not required.
Please see https://github.com/openshift/openshift-docs/pull/75714#discussion_r1619266559

----
+

> [NOTE]

Reviewer: If the suggested changes on lines 70 and 74 are acceptable, then we can remove this note.

Author: This note is required.
Is the phrasing correct?

Reviewer: The note placement seems weird. Is it possible to add an actual step instead instead of an admonition? Something like:
`.. Log in to the guest as a root user by using the following command:`

Author: There are at least thee routes a user could take:
1. Login directly as root (this is somethings disabled).
2. Login as a regular user and switching to root (using `su` or `sudo su`).
3. Login as a regular user and executing each command using `sudo`.

This is why I think we should not limit the instructions to use a single command.
WDYT?

Reviewer: For the purpose of this procedure, wouldn't option 2 make the most sense? Especially because all the commands after `virtctl console <vm_name>` have the `#` prompt which indicates a root user. How about `.. If you are not already logged in as a root user, switch to the root user account to execute the remaining configuration steps.`

I am not completely opposed to a Note, but we would need to change the indentation to left so that it doesn't seem like it applies to the previous step (visually). Adjusting the indentation might cause the list numbering to break.

Author: Done.

====
The following commands should all run as the root user.
====
.. Disable the `irqbalance` service by using the following command:
+
[source,terminal]
----
# systemctl disable irqbalance && systemctl stop irqbalance
----

> .. Enable the realtime and nvf repos:

Reviewer: ```suggestion
.. Enable the real-time and network function virtualization (NFV) repositories:
```

+
[source,conf]
----
#
# Variable settings below override the definitions from the
# /etc/tuned/realtime-variables.conf file.

#
# Core isolation
#
# The 'isolated_cores=' variable below controls which cores should be
# isolated. By default we reserve 1 core per socket for housekeeping
# and isolate the rest. But you can isolate any range as shown in the
# examples below. Just remember to keep only one isolated_cores= line.
#
# Examples:
# isolated_cores=2,4-7
# isolated_cores=2-23
#
# Reserve 1 core per socket for housekeeping, isolate the rest.
# Change this for a core list or range as shown above.
isolated_cores=2-3 <1>

#
# Uncomment the 'isolate_managed_irq=Y' bellow if you want to move kernel
# managed IRQs out of isolated cores. Note that this requires kernel
# support. Please only specify this parameter if you are sure that the
# kernel supports it.
#
isolate_managed_irq=Y <2>

#
# Set the desired combined queue count value using the parameter provided
# below. Ideally this should be set to the number of housekeeping CPUs i.e.,
# in the example given below it is assumed that the system has 4 housekeeping
# (non-isolated) CPUs.
#
# netdev_queue_count=4
----

> <1> The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.

Reviewer: ```suggestion
<1> Set the first two CPUs (0 and 1) aside for house keeping tasks and isolate the rest for the real-time application.
```

Author: We actually set the isolated CPU, and the reserved are the ones left.
Is there an alternative phrasing you think we can use here?

> <2> isolate_managed_irq should be set to `Y`

Reviewer: ```suggestion
<2> Uncomment the `isolate_managed_irq=`Y` line to move kernel-managed interrupt requests out of isolated cores.
```

Author: This line is not commented by default.
Is there an alternative phrasing we could use?

Reviewer: Maybe change it to:
```suggestion
<2> Set the `isolate_managed_irq` parameter to `Y` to move kernel-managed interrupt requests out of isolated cores.
```

Author: Done.

----

.. Make sure irqbalance is disabled:
+
[source,terminal]
----
$ systemctl disable irqbalance && systemctl stop irqbalance
----

.. Install the following packages:
+
[source,terminal]
----
$ dnf install kernel-rt tuned cloud-init realtime-tests
----

.. To achieve low-latency tuning by using the `realtime-virtual-guest` profile in the TuneD application, run the following commands:
+
[source,terminal]
----
$ dnf --enablerepo=rt install -y tuned-profiles-realtime
$ dnf --enablerepo=nfv install -y tuned-profiles-nfv-guest
----

.. Set isolated_cores=2-3 in /etc/tuned/realtime-virtual-guest-variables.conf
.. Make sure isolate_managed_irq=Y is set in /etc/tuned/realtime-virtual-guest-variables.conf

The first two CPUs (0 and 1) are set aside for house keeping tasks and the rest are isolated for the real-time application.
+
[source,terminal]
----
$ tuned-adm profile realtime-virtual-guest
----

.. Set up the real-time kernel to be the default:
+
[source,terminal]
----
$ grubby --set-default=$(ls /boot/vmlinuz* | grep rt)
$ grubby --args="iommu=pt intel_iommu=on default_hugepagesz=1G idle=poll" --update-kernel=$(grubby --default-kernel)
----

. Restart the VM to apply the changes.

> . Test the guest latencies

Reviewer: I would split this part into a separate `Verification` section

```suggestion
.Verification

. Use the `cyclictest` tool to verify that the real-time guest is configured properly:
```

+
[source,terminal]
----
$ cyclictest --priority 1 --policy fifo -h 50 -a 2-3 --mainaffinity 0,1 -t 4 -m -q -i 200 -D 12h
----

> After the test finishes, we can get the “Max Latencies” in the output, it should be less than 40 micro seconds.

Reviewer: Is there an example output that we can add here?

$ oc apply -f <file_name>.yaml
----

. Configure the guest operating system. The following example shows the configuration steps for a {op-system-base} 9 operating system:
.. Run the following command to connect to the VM console:
+
[source,terminal]
----
$ virtctl console <vm_name>
----

.. Make sure irqbalance is disabled:
+
[source,terminal]
----
$ systemctl disable irqbalance && systemctl stop irqbalance
----

.. Install the following packages:
+
[source,terminal]
----

> $ dnf install kernel-rt tuned cloud-init rt-tests

Author: Thank you, changed to `realtime-tests`.

----

.. To achieve low-latency tuning by using the `realtime-virtual-guest` profile in the TuneD application, run the following commands:
+
[source,terminal]
----
$ dnf --enablerepo=rt install -y tuned-profiles-realtime
$ dnf --enablerepo=nfv install -y tuned-profiles-nfv-guest
----
+
[source,terminal]
----

> $ echo isolated_cores=2-3 > /etc/tuned/realtime-virtual-guest-variables.conf

Reviewer 4: Maybe a typo, in following cyclictest command, '-a 2-4' is used, while the isolated cores are 2-3 here.

Author: Thank you, changed to `2-3`.

> $ echo isolate_managed_irq=Y >> /etc/tuned/realtime-virtual-guest-variables.conf

Reviewer 4: Usually this setting has already been in the conf file, maybe it's better to have a check before setting it.
